### DISCUSSION TITLE
Can libvips check if a tif or psd file has layers and write each layer as a separate file?

### DISCUSSION THREAD

**Discussion by hanssonrickard** [libvips/libvips#2175] [Q&A]
Hi.
Can libvips check if a tif or psd file has layers and if so, write each layer to a separate output file?

I tried vipsheader -a but it shows no info about possible layers in the example tif file (contains two layers).

**Comment by jcupitt** [comment]
Hi @hanssonrickard, I'm not sure, could you post a sample tiff?

**Reply by hanssonrickard** [reply]
Sure. 
It is just a sample file with three layers and somthing drawn och each layers.

[transparent_and_layers.tif.zip](https://github.com/libvips/libvips/files/6237997/transparent_and_layers.tif.zip)

**Comment by jcupitt** [comment]
I'm afraid your TIFF is just a simple RGB image, there are no layers. Photoshop has hidden the extra layer info in a combination of XML metadata in the photoshop-data tag and a proprietary 0x935c tag.

It looks like a regular TIFF, but it isn't. You'll need a specialist library to decode this format.

**Reply by hanssonrickard** [reply]
Ok.
Thanks for taking the time to look at it.

I will then have to fallback to imagemagick for those files that contains layers.
I can use imagemagicks "identify" to get info about possible layers:

identify -format "%[scene] " /CAVOKDATA/CUSTOMERS/Demokund/Rickard/transparent_and_layers.tif 

Which, for my sample file will give following output: "0 1 2 3"
But for a file that does not have any layers, i will get value "0" only.

Based on that i will then have to do the conversion with Imagemagick which takes the layer id as part of the inputname.
Here is a bash example:

LAYERS=`$TIMEOUT $IDENTIFY -format "%[scene] " "$FILEPATH"`
for LAYER in $LAYERS; do
           convert "$FILEPATH[0]" "$FILEPATH[$LAYER]" \( -clone 0 -alpha transparent \) -swap 0 +delete -coalesce -compose src-over -composite -resize ${PreviewSize}x${PreviewSize} -size ${PreviewSize}x${PreviewSize} "${BACKGROUND_OPTIONS}" $BACKGROUND_FILL -compose dstover -composite -intent relative $PROFILES $ROTATE -quality 70 "$TMPDIR/$$-${LAYER}.jpg"

This will give me four files, one composite (0) and three layer versions (1, 2, 3).
